book/src/rust/rust_style_guide.md

```diff
@@ -0,0 +1,660 @@
+pub mod feature_gated_public_module;
+```
+
+#### Imports and Exports
+
+Imports (use) and exports (mod) should be ordered as follows:
+
+1. External Crate Declarations
+
+2. Private Imports
+
+3. Private Imports with Features
+
+4. Private Exports
+
+5. Private Exports with Features
+
+6. Public Exports
+
+7. Public Exports with Features
+
+8. Reexports
+
+9. Reexports with Features
+
+Here's an example set of declarations:
+
+```rust
+extern crate crate_name;
+
+use module::submodule::entry;
+
+#[cfg(feature = "...")]
+use module::feature_gated_submodule;
+
+mod another_module;
+mod module;
+mod the_third_module;
+
+#[cfg(feature = "...")]
+mod feature_gated_module;
+
+pub mod public_module;
+
+#[cfg(feature = "...")]
+#[cfg_attr(doc_cfg, doc(cfg(feature = "...")))]
+pub mod feature_gated_public_module;
+
+pub use reexported_objects;
+
+#[cfg(feature = "...")]
+#[cfg_attr(doc_cfg, doc(cfg(feature = "...")))]
+pub use feature_gated_reexported_objects;
+```
+
+Ensure that there are new lines between each category. Be sure that if some imports or exports are feature-gated, that they are sorted by feature alphabetically. If there is a feature gated import that requires importing multiple objects, use the following pattern:
+
+```rust
+#[cfg(feature = "...")]
+use {
+    thing1, thing2, thing3, thing4,
+};
+```
+
+**NOTE**: All imports should occur at the top of any module, and a newline should be added between the last import and the first declared object.
+
+#### Let binding
+
+
+##### Always separately bind [RAII](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization) guards
+
+
+>RAII stands for "Resource Acquisition is Initialisation". The essence of the pattern is that resource initialisation is done in the constructor of an object and finalisation in the destructor. This pattern is extended in Rust by using a RAII object as a guard of some resource and relying on the type system to ensure that access is always mediated by the guard object.
+
+Prefer
+
+```rust
+fn use_mutex(m: sync::mutex::Mutex<int>) {
+    let guard = m.lock();
+    do_work(guard);
+    drop(guard); // unlock the lock
+    // do other work
+}
+```
+over
+
+```rust
+fn use_mutex(m: sync::mutex::Mutex<int>) {
+    do_work(m.lock());
+    // do other work
+}
+```
+
+##### Prefer conditional expressions to deferred initialization.
+
+Prefer
+
+```rust
+let foo = match bar {
+    Baz  => 0,
+    Quux => 1
+};
+```
+
+Over
+
+```rust
+let foo;
+match bar {
+    Baz  => {
+        foo = 0;
+    }
+    Quux => {
+        foo = 1;
+    }
+}
+```
+
+unless the conditions for initialization are too complex to fit into a simple conditional expression.
+
+##### Use type annotations for clarification; prefer explicit generics when inference fails.
+
+Prefer
+
+```rust
+s.iter().map(|x| x * 2)
+        .collect::<Vec<_>>()
+```
+
+Over
+
+```rust
+let v: Vec<_> = s.iter().map(|x| x * 2)
+                        .collect();
+```
```

Comment:
I actually have the opposite preference, but I don't think one is *better* than the other. IMO we shouldn't have a rule for this.

Also it's not always possible to specify generics at the function call-site, so we'd still end up with inconsistencies

Reply:
I prefer the second as well,  as its more "natural" to people coming from other languages.  I think the goal should be clarity to the human reading it (which is subjective), and in this case it depends too much on the actual code to say what would be "clearer".  But in this case, I think the second would be clearer to more people.

I would prefer if this rule was more about documentation, because these constructs quickly become difficult to reason through.  So, I would prefer the rule say,  "these kinds of constructs should also ALWAYS be accompanied by a comment explaining WHAT it is doing.

```rust
// v becomes a vector where every element of s is doubled.
let v: Vec<_> = s.iter().map(|x| x * 2)
                        .collect();
```